server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -230,40 +225,41 @@ private void getFromTranslog(
         ClusterStateObserver observer,
         ActionListener<GetResponse> listener
     ) {
-        tryGetFromTranslog(request, indexShard, state, listener.delegateResponse((l, e) -> {
-            final var cause = ExceptionsHelper.unwrapCause(e);
-            logger.debug("get_from_translog failed", cause);
-            if (cause instanceof ShardNotFoundException || cause instanceof IndexNotFoundException) {
-                logger.debug("retrying get_from_translog");
-                observer.waitForNextChange(new ClusterStateObserver.Listener() {
-                    @Override
-                    public void onNewClusterState(ClusterState state) {
-                        getFromTranslog(request, indexShard, state, observer, l);
-                    }
+        try {
+            final var node = getCurrentNodeOfPrimary(state, indexShard.shardId());
+            final var retryingListener = listener.delegateResponse((l, e) -> {
```

Comment:
@henningandersen Does it really matter? i preferred it after `getCurrentNodeOfPrimary` since if that throws, we don't need the `retryingListener`. I've moved it, but I want to know why it is preferred. From what I've seen in the code base, it is actually more common to define listeners like this right in the `tryGetFromTranslog` method call, which would be like what I had.

Reply:
sorry, I were trying to multi-line comment, apparently I failed? It should also concern the `tryGetFromtTranslog` method.

That method has an `ActionListener` and thus should report any exceptions through that, hence there is no need for it to be wrapped. We could assert that it does not throw, but I found that excessive. Hence I'd prefer to just move it out of the try/catch.

Reply:
Ok, so it seems you meant to move out all except `getCurrentNodeOfPrimary`. Sure.

Reply:
done in https://github.com/elastic/elasticsearch/pull/105028/commits/15439c6dedda727f724b460721dc59a9524cd28d.